### ISSUE TITLE
Spike: create Instructor Email Digests architecture and prototype

### ISSUE THREAD

**Issue by seanh** [hypothesis/lms#5075]
Create a working local prototype of the Instructor Email Digests MVP (https://github.com/hypothesis/lms/issues/4954), and based on that prototype write up a plan for how we're going to architect the feature.

**Comment by seanh** [comment]
# The Plan

Here's how we're currently planning to implement the email digests feature, based on the prototype:

## New "activity" API in h

We need to add a service to h that reports annotation counts for a given time period grouped by group and user. The prototype uses an API that we might call an "annotation activity API" or "counts API" or "stats API". It accepts a `user`, `since` (date string) and `until` date string as query params and the response bodies look something like this:

```json
{
  "groups": [
    {
      "authority_provided_id": "<AUTHORITY_PROVIDED_ID>,
      "users": [
        {
          "userid": "acct:<USERNAME>@<AUTHORITY>",
          "num_annotations": 42,
        },
        {
          "userid": "acct:<USERNAME>@<AUTHORITY>",
          "num_annotations": 24,
        },
        ...
      ],
    },
    {
      "authority_provided_id": "<AUTHORITY_PROVIDED_ID>,
      "users": [ ... ]
    },
    ...
}

###### Alternatives

It was [suggested in Slack](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1676553161695279?thread_ts=1675861694.508839&cid=C1MA4E9B9) that we could instead implement a "list annotations" API that returns a list of annotation dicts filtered according to some query params:

```json
[
    {"group": "group_id_123", "user": "user_id_1"},
    {"group": "group_id_123", "user": "user_id_1"},
    {"group": "group_id_123", "user": "user_id_2"},
    {"group": "group_id_123", "user": "user_id_2"},
    {"group": "group_id_123", "user": "user_id_2"},
    ...
]
```

The LMS app's Python code would then iterate over the list of annotations and build up the per-group-per-user counts. Pagination may unfortunately be required but we might be able to get away without it for the first version. This may be more general compared to the "annotation counts" API (which feels a little contrived) and may also be more flexible. Currently [the h API](https://h.readthedocs.io/en/latest/api-reference/v2/) has a way to fetch a single annotation (`GET /api/annotations/<id>`) but it doesn't have a way to fetch a list of annotatins (`GET /api/annotations`) except for the search endpoint. So this may fill a gap in the existing API.

###### Pull requests

- [ ] Add either an "annotation counts" or a "list annotations" service and API to h

## New `UserSettings` model

We need to store an `instructor_email_digests` setting against users. This can't just be added to the `public.users` table in the LMS app (`lms.models.User`) because that table is grouped by application instance so it can contain multiple rows for the same user if an LMS has multiple application instances.

The prototype adds a new `lms.models.UserSettings` model to allow us to store user settings that aren't grouped by application instance.

###### Alternatives

We might want to think of the new table as just a second user table, this one not grouped by application instance, rather than thinking of it specifically as a "user settings" table. This may involve moving the existing [`HUser` model](https://github.com/hypothesis/lms/blob/main/lms/models/h_user.py) into the DB?

###### Pull requests

- [ ] DB migration to add `user_settings` table
- [ ] Add `UserSettings` model

## Track LTI roles against courses, not assignments

The same user may be an instructor in one course and a learner in another course within the same LMS. Email digests should only cover activity in courses in which the user is an instructor. So once the LMS app has grouped the user's annotation counts by course, for each course it needs to ask _Is this user an instructor in this course?_

The prototype does this by adding a `GroupingMembership.lti_role_id` column and `GroupingMembership.lti_role` relationship and updating `GroupingService.upsert_grouping_memberships()` to populate this column.

###### Considerations

LTI roles only make sense against course groupings so we may want to add a Postgres `CHECK` constraint that `grouping_membership.lti_role_id` must be `NULL` for all non-couse groupings ([slack thread](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1676892516716849)).

###### Alternatives

The code currently [tracks LTI roles against assignments](https://github.com/hypothesis/lms/blob/1b45267b7656dd863bacb83320747a32deec81b7/lms/models/assignment_membership.py#L26-L30). Instead of adding a new column we could make use of this existing column by saying that a user is an instructor in a course if they're an instructor in any assignment in that course. But tracking roles against assignments is fundamentally invalid: LTI roles applies to courses ("contexts") not assignments, and tracking roles against assignments means that the same user could have different roles in different assignments in the same course.

###### Pull requests

- [ ] DB migration to add `grouping_membership.lti_role_id` column and `CHECK` constraint
- [ ] Model code to add `GroupingMembership.lti_role_id` column, `GroupingMembership.lti_role` relationship, and `CHECK` constraint
- [ ] Update `GroupingService.upsert_grouping_memberships()` to populate the new column

## Enable the `instructor_email_digests` preference for instructors at participating institutions

Our intention is to automatically opt-in all instructors at participating institutions to receiving email digests about activity in all assignments in all courses in which they're instructors.

The way the prototype does this is that when a user launches an assignment it checks whether the `ApplicationInstance`-level `instructor_email_digests` feature flag is on and if so it checks whether the user is an instructor (`request.user.is_instructor`) and if so it sets the user's `instructor_email_digests` preference to `True`.

This means that instructors won't actually start getting emails until they launch one assignment (in any course).

We could also run some sort of one-off DB migration or task to turn on the `instructor_email_digests` preference for all instructors at participating institutions. If there's an admin page for turning on the feature flag for an `ApplicationInstance` that could also turn on the preference for all of that `ApplicationInstance`'s instructors. But note that we'd still need the code for turning on the preference when an instructor does a launch, because this will still be needed for new instructors.

###### Pull requests

- [ ] Turn on the `instructor_email_digests` user preference when an instructor launches an assignment

## New `EmailDigestsService` in LMS

This is a service that produces the activity data (counts of annotations in courses) that goes into email digests. This service doesn't actually send the emails (that's done by a Celery task, see below).

This service calls the new h API to get the activity for a given set of users in a given time period. While the h API speaks in terms of h concepts like annotations, user, and groups, this service combines the h API's data with data from LMS's DB about instructors, courses, etc. The service handles:

* Rolling up activity by course: figuring out which course each h group belongs to and summing up the group annotation counts to make course annotation counts
* Not counting annotations in groups that belong to courses in which the user isn't an instructor
* Not counting the user's own annotations

###### Pull requests

- [ ] Add the new `EmailDigestsService` to LMS

## New `MAILCHIMP_API_KEY` envvar / `mailchimp_api_key` Pyramid setting

###### Pull requests

- [ ] Add `mailchimp_api_key` setting

## New `mailchimp-transactional` Python requirement

[mailchimp-transactional](https://pypi.org/project/mailchimp-transactional/) is the Mailchimp Transactional API's official Python client, which we intend to use for actually sending the emails. ([source code on GitHub](https://github.com/mailchimp/mailchimp-transactional-python), [docs](https://mailchimp.com/developer/transactional/api/).

###### Pull requests

- [ ] Add `mailchimp-transactional` Python requirement to LMS

## New `send_email_digests()` Celery task

There needs to be a `send_email_digests()` Celery task in the LMS app that gets a given user's activity for "yesterday" (5AM to 5AM UTC) from the LMS app's `EmailDigestsService` and calls the Mailchimp Transactional API to send that user's email digest.

In the prototype the `send_email_digests()` task handles one user per task run, but it's probably worth making it handle `n` users per task run instead as this may be more efficient (in particular it may reduce the number of calls to the h API).

###### Pull requests

- [ ] Add `send_email_digests()` Celery task to LMS

## New `schedule_email_digests()` Celery task

There needs to be a scheduled task in the LMS app that runs once per day (shortly after 5am UTC) and that finds all the instructors who we need to send email digests to and queues a `send_email_digests()` task for each user.

###### Pull requests

- [ ] Add `schedule_email_digests()` Celery task to LMS
- [ ] Add code to the LMS beat in h-periodic to call the `schedule_email_digests()` task each day at 5AM UTC

## Unsubscribing from emails

**Comment by seanh** [comment]
Notes from the backend team meeting about this (I'll integrate these into the plan above later):

* We'll go with the "list annotations"-style API in h. We realised that there's some potential confusion with a `filter[users]` query param to this API in that it's not filtering to annotations by the given users but rather to annotations in groups that the given users are members of. So we may just need a different name for the param that makes its meaning clear. Or we may want a `filter[groups]` param instead and require the LMS app to find the list of groups that the users belong to, but that may lead to really long query params if there are a lot of groups. I'll create a separate ticket tomorrow where we can discuss the API design anyway.

* We'll just stick with `ApplicationInstance`-based settings, as we don't want to bite off figuring out how `Organization` settings should work. In practice we'll have to decide what it means if a user belongs to multiple `ApplicationInstance`'s that have conflicting settings. The suggestion was that if the setting is enabled for any of the user's `ApplicationInstance`'s then we consider it enabled for that user. (This is similar to the prototype which toggles on the user preference when the user launches an assignment with an enabled `ApplicationInstance`. The difference is that with the prototype each user has to actually do a launch with an enabled `ApplicationInstance` in order to trigger it, which is actually probably undesirable because it means that the behaviour depends on what launches each user has done.)

* It sounds like we want to try to get away without storing any per-user setting in the DB, and instead the Celery task will just find all instructors in enabled `ApplicationInstance`'s. Need to figure out what that DB query would be and if it's not too hard to write, sounds fine. I think we'll probably need to move to per-user settings straight away after releasing the first version because we can't continue to rely on Mailchimp unsubscribes because they expire after a year (which also means we'll have to implement our own unsubscribe links).

* It sounds like we want to add a new user table that is not divvied-up by `ApplicationInstance`. We talked about having `user` and `instance_user` tables, but that will require renaming the existing `user` table to `instance_user`. Once we have these two tables I suspect we may no longer need the non-DB `HUser` model. I think I may want to find a reasonably quick and easy solution here without biting too much off, as it sounds like there could be a fair bit of refactoring here.

* We may want to look into the current assignment-based user roles in the DB and see if there are any cases where a user has different roles in different assignments within the same course. I suspect there totally could be though because this would happen even if a user's role in a course changes and then they launch an assignment with their new role but don't go back and re-launch a previously-launched assignment.

* We talked about extending either the `UserService` or (more likely) the `LTIRoleService` to have methods for returning the list of all instructors or querying whether a given user is an instructor in a given course, etc. That would mean that this service can wrap any considerations about whether LTI roles are looked up by course (which would require adding a new DB column) or by assignment.

* We'd like the code for actually talking to the Mailchimp API (or calling their Python client) to be in a service, and keep the Celery tasks as thin as possible. I think for practicality this will probably just be a non-generalized `EmailDigestsService` for now and parts of it can be generalized for reuse in future if we ever have a need to reuse them. For unit testing purposes some of the service's code will probably be split out into a helper.

* We will use the `mailchimp-transactional` Python package even though it looks like a low quality Python package because it is the official client and it means that we should get Dependabot updates for any Mailchimp API changes.
